### DISCUSSION TITLE
payload.db.drizzle types

### DISCUSSION THREAD

**Discussion by xantiagoma** [payloadcms/payload#4318] [Roadmap]
Hi everyone,

Love the changes that have been introduced to the framework with the Drizzle ORM / Postgres integration.

I think It has many potential as a schema builder also,
I'm aware that with Local API you can use for example `payload.find({'collection': 'users'});` on the new docs there's a mention to [accessing drizzle](https://payloadcms.com/docs/database/postgres#access-to-drizzle) it works great but there's a lack of types definition on this object as it is typed as `export type DrizzleDB = NodePgDatabase<Record<string, unknown>>;`

it would be great if it is typed (automatically or as part of `"generate:types"` command) so editor IntelliSense can for example auto-complete statements as `payload.db.drizzle.query.users.findFirst()` or `payload.db.drizzle.select(payload.db.drizzle._.schema.posts_locales)` etc...

**Comment by jmikrut** [comment]
This is planned actually! Thanks for bringing this up. Converting to a roadmap discussion and marking as priority 1.

Expect this to come soon!

PS - someone in our Discord already figured out how to do this!

https://discord.com/channels/967097582721572934/967097582721572937/1161393807414268025

**Reply by willviles** [reply]
This is awesome news ❤️ 

Of course, working with the underlying drizzle instance will have the full DX, and that's great.

But even more importantly, it will now be possible to generate zod schemas using [drizzle-zod](https://orm.drizzle.team/docs/zod). 

Those of us who use trpc with zod schema validation will get a massive productivity boost!

**Reply by isaiahgrey93** [reply]
@jmikrut this discord link no longer seems to be valid? Is there detail on this somewhere else?

**Comment by ericwaetke** [comment]
Would it be possible to export/expose this in a monorepo for example? That way we’d have the advantage of querying the db directly without any http requests from any app using drizzle, right?

**Comment by valerioleo** [comment]
I just tried the new v3 (massive achievement, congrats 🎉) where I tried to get a typed `payload.db.drizzle` or even `payload.db.(tables|relationships|enums)` without success. 

The new drizzle hooks unlock a lot on new possibilities, but it depends a lot from having a typed schema in payload. 
Is this something that can be achieved in v3?

**Comment by Robstei** [comment]
I faced the same problem and found an [answer on Discord](https://discord.com/channels/967097582721572934/967097582721572937/1310357295749795943) (thanks ritsu).

> Try this:
> ```
> import type * as  schema from './introspect'
> import type { NodePgDatabase } from 'drizzle-orm/node-postgres'
> export const getDB = (payload) => payload.db.drizzle as NodePgDatabase<typeof schema>
> 
> // db is typed
> const db = getDB(payload) 
> ``` 
> where ./introspect is your file with the introspected schema
> 

He also mentioned that this might be integrated into payload in the future.

**Comment by r1tsuu** [comment]
PR with this is here! https://github.com/payloadcms/payload/pull/9953
